### DISCUSSION TITLE
Does SwiftUSD use Rust OpenUSD bindings?

### DISCUSSION THREAD

**Discussion by beeeeep54** [wabiverse/swift-usd#21] [Q&A] [answered]
Hi 👋

I stumbled across this project when researching OpenUSD bindings for Rust, and I noticed that this project uses Rust as well. Poking around in the code appeared to show some kind of Rust Swift interop, but I can't find where the actual C++ interop is?
  
The main thing I'm wondering is, does this project use OpenUSD Rust bindings? And if not, what does it use?

Thanks!

**Comment by furby-tm** [comment, marked as answer]
Hey there @beeeeep54! This repo does not currently offer any Rust/C++ interop, the code found [here](https://github.com/wabiverse/SwiftUSD/blob/main/Sources/Rust/main.rs#L14-L19) is some experimentation with Swift/Rust interop, and that example is only related to one of OpenUSD's dependencies **Ptex**, that is - [this line of Swift, calling directly into (**Ptex**) C++ code](https://github.com/wabiverse/SwiftUSD/blob/main/Sources/rust-usd/rust-usd.swift#L62) being called from Rust through [swift-bridge](https://github.com/chinedufn/swift-bridge) to convert a Rust `u32` unsigned integer to it's equivalent `Ptex.MeshType` enum value, more as a "simple, dumb, test".

If you're interested in **OpenUSD Rust bindings** you can find that [here](https://github.com/vfx-rs/usd-bind), @scott-wilson is the best resource for more information on that, and you can find him on the [ASWF slack rust channel](https://academysoftwarefdn.slack.com/archives/C01CJ73NQ79).

If you're interested in how interoperability is working between Swift and OpenUSD's C++ codebase, [here](https://www.swift.org/documentation/cxx-interop/) is how Swift/C++ interop works.

### Steps to do this easily...

1. Create a new swift package by running the following in your terminal.
```swift
mkdir MyApp
cd MyApp

swift package init
```

2. Create a new directory, say `MyCxx` inside the `MyApp/Sources` directory with the following structure.
```swift
Sources/
  MyCxx/
    include/
      MyCxx.h
    MyCxx.cpp
```

3. Inside the `Package.swift` file at the project's root, copy the following:
```swift
// swift-tools-version: 5.10
import PackageDescription

let package = Package(
  name: "MyApp",
  platforms: [
    .macOS(.v14),
    .visionOS(.v1),
    .iOS(.v17),
    .tvOS(.v17),
    .watchOS(.v10)
  ],
  targets: [
    .target(
      name: "MyCxx"
    ),

    .executableTarget(
      name: "MyApp",
      dependencies: [
        .target(name: "MyCxx")
      ],
      swiftSettings: [
        // enable c++ interop.
        .interoperabilityMode(.Cxx)
      ]
    )
  ],
  // set c++ language standard.
  cxxLanguageStandard: .cxx17
)
```

4. Inside `MyApp/Sources/MyCxx` add some simple c++ code to test with.
```cpp
// MyCxx.h
#pragma once

void doSomething();
```
```cpp
// MyCxx.cpp
#include "MyCxx.h"

#include <stdio.h>

void doSomething()
{
  printf("Hello from C++!\n");
}
```

5. Inside `MyApp/Sources/MyApp/MyApp.swift` add the following to interop with your c++ library directly in swift.
```swift
import MyCxx

@main
struct MyApp {
  static func main() {
    doSomething() // should print "Hello from C++!"
  }
}
```

6. Back in your terminal, build and run your Swift/C++ app.
```pwsh
swift run
# "Hello from C++!"
```

Hopefully that helps! Cheers 👋🏻.

**Comment by beeeeep54** [comment]
Thanks so much for the help!

**Reply by furby-tm** [reply]
Anytime! Feel free to reach out about anything else 👍🏻.
